builtin/mainmenu/dlg_mg_flags.lua

```diff
@@ -0,0 +1,211 @@
+			else
+				flag_list[text[i]] = true
+			end
+		end
+	end
+	mg_flags_helper(flag_list_g, core.setting_get("mg_flags"))
+	mg_flags_helper(flag_list_6, core.setting_get("mgv6_spflags"))
+	mg_flags_helper(flag_list_7, core.setting_get("mgv7_spflags"))
+	
+	water_level = core.setting_get("water_level")
+	
+	update_mg_flags()
+end
+
+local function mg_flags_formspec(dialogdata)
+
+	-- load minetest.conf for first time
+	if mg_first_time == true then
+		load_mg_flags()
+		update_mg_flags()
+	end
+	mg_first_time = false
+	
+	-- this is it!
+	local formspec =
+		"size[8,4,true]" ..
+		"button[4,3.65;3,0.5;mg_ok;" .. fgettext("Ok") .. "]" ..
+		
+	-- special mapgen-specific flags
+		"label[0,0;" .. core.setting_get("mg_name") .. " " .. fgettext("Mapgen flags") .. "]"
+	if core.setting_get("mg_name") == "v5" then
+		formspec = formspec ..
+			"label[0,1.0;" .. fgettext("No flags in v5 mapgen") .. "]"
+	elseif core.setting_get("mg_name") == "v6" then
+		formspec = formspec ..
+			"checkbox[0.25,0.5;cb_biomeblend;" .. fgettext("Biome blend") .. ";" ..
+				dump(flag_list_6["biomeblend"]) .. "]" ..
+			"checkbox[0.25,1.0;cb_jungles;" .. fgettext("Jungles") .. ";" ..
+				dump(flag_list_6["jungles"]) .. "]" ..
+			"checkbox[0.25,1.5;cb_snowbiomes;" .. fgettext("Snowbiomes") .. ";" ..
+				dump(flag_list_6["snowbiomes"]) .. "]" ..
+			"checkbox[0.25,2.0;cb_mudflow;" .. fgettext("Mud flow") .. ";" ..
+				dump(flag_list_6["mudflow"]) .. "]"
+	elseif core.setting_get("mg_name") == "v7" then
+		formspec = formspec ..
+			"checkbox[0.25,0.5;cb_mountains;" .. fgettext("Mountains") .. ";" ..
+				dump(flag_list_7["mountains"]) .. "]" ..
+			"checkbox[0.25,1.0;cb_ridges;" .. fgettext("Ridges") .. ";" ..
+				dump(flag_list_7["ridges"]) .. "]"
+	elseif core.setting_get("mg_name") == "singlenode" then
+		formspec = formspec ..
+			"label[0,1.0;" .. fgettext("No flags in singlenode mapgen") .. "]"
+	end
+	
+	-- general attributes a.k.a. generated structure
+	formspec = formspec ..
+		"label[4,0;" .. fgettext("Generated structure") .. "]" ..
+		-- trees, caves, flat, dungeons, light.
+		"checkbox[4.25,0.5;cb_trees;" .. fgettext("Trees") .. ";" ..
+				dump(flag_list_g["trees"]) .. "]" ..
+		"checkbox[4.25,1.0;cb_caves;" .. fgettext("Caves") .. ";" ..
+				dump(flag_list_g["caves"]) .. "]" ..
+		"checkbox[4.25,1.5;cb_flat;" .. fgettext("Flat world") .. ";" ..
+				dump(flag_list_g["flat"]) .. "]" ..
+		"checkbox[4.25,2.0;cb_dungeons;" .. fgettext("Dungeons") .. ";" ..
+				dump(flag_list_g["dungeons"]) .. "]" ..
+		"checkbox[4.25,2.5;cb_light;" .. fgettext("Sunlight") .. ";" ..
+				dump(flag_list_g["light"]) .. "]"
+	
+	-- water level
+	formspec = formspec ..
+		" field[0.25,3.9;3.75,0.5;te_water_level;" .. fgettext("Water level") .. ";" .. water_level .. "] "
+	
+	return formspec
+end
+
+local function mg_flags_buttonhandler(this, fields)
+
+	if fields["mg_ok"] then
+		update_mg_flags()
+		this:delete()
+		return true
+	end
+	
+	function helper_func(flag_list)
```

Comment:
I've found an even more compact way of writing this. Instead of a function you do an outer for loop over a previously declared `local flag_lists = {flag_list_g, flag_list_6, flag_list_7}`